nixos/modules/services/security/endlessh-go.nix

```diff
@@ -0,0 +1,96 @@
+{ config
+, lib
+, pkgs
+, ...
+}:
+
+with lib;
+
+let
+  cfg = config.services.endlessh-go;
+in
+{
+  options.services.endless-go = {
+    enable = mkEnableOption "Whether to enable the endlessh-go service.";
+
+    listenAddress = mkOption {
+      type = types.str;
+      default = "0.0.0.0";
+      description = mdDoc ''
+        Interface address to bind the endlessh-go daemon to SSH connections.
+      '';
+    };
+
+    port = mkOption {
+      type = types.port;
+      default = 2222;
+      description = mdDoc ''
+        Specifies on which port the endlessh-go daemon listens for SSH
+        connections.
+      '';
+    };
+
+    prometheus = {
+      enable = mkEnableOption "Whether to enable Prometheus integration.";
+
+      listenAddress = mkOption {
+        type = types.str;
+        default = "0.0.0.0";
+        description = mdDoc ''
+          Interface address to bind the endlessh-go daemon to answer Prometheus
+          queries.
+        '';
+      };
+
+      port = mkOption {
+        type = types.port;
+        default = 2112;
+        description = mdDoc ''
+          Specifies on which port the endlessh-go daemon answers for Prometheus
+          queries.
+        '';
+      };
+    };
+
+    extraOptions = mkOption {
+      type = with types; nullOr (listOf str);
+      default = [ ];
+      description = mdDoc ''
+        Additional command line options to pass to the endlessh-go daemon.
+      '';
+    };
+
+    openFirewall = mkEnableOption ''
+      Whether to open firewall port for the SSH listener.
+    '';
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.endlessh-go = {
+      description = "SSH tarpit";
+      requires = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        Restart = "always";
+        ExecStart = with cfg; concatStringsSep " " ([
+          "${pkgs.endlessh-go}/bin/endlessh-go"
+          "-logtostderr"
+          "-host=${listenAddress}"
+          "-port=${toString port}"
+        ] ++ optionals prometheus.enable [
+          "-enable_prometheus"
+          "-prometheus_host=${prometheus.listenAddress}"
+          "-prometheus_port=${toString prometheus.port}"
+        ] ++ optionals extraOptions);
+        KillSignal = "SIGTERM";
+        AmbientCapabilities = "CAP_NET_BIND_SERVICE";
+        DynamicUser = true;
```

Comment:
that's a good start for hardening, but there are probably a lot more options that could be set. would make sense to do so as this *is* going to be a service under attack.

Reply:
I've added a couple of options but I still think that this is not enough. Could you please suggest a service which I can use as an example?

Reply:
services usually don't make for good examples, but you can run `systemd-analyze security endlessh-go` and pick suggestions from there until the score is low